File: leetcode/python/435_non-overlapping_intervals.py
from bisect import bisect_left

class Solution:
    def eraseOverlapIntervals(self, intervals: list[list[int]]) -> int:
        intervals.sort()
        n = len(intervals)
        dp = [0] * (n+1)
        dp[-2] = 1
        starts = [e[0] for e in intervals]
        for i in range(n-2, -1, -1):
            j = bisect_left(starts, intervals[i][1])
            dp[i] = max(dp[i+1], 1+dp[j])
        return n - dp[0]

s = Solution()
print(s.eraseOverlapIntervals([[1,2],[2,3],[3,4],[1,3]]))
print(s.eraseOverlapIntervals([[1,2],[1,2],[1,2]]))
print(s.eraseOverlapIntervals([[1,2],[2,3]]))
print(s.eraseOverlapIntervals([[0,2],[1,3],[2,4],[3,5],[4,6]]))

File: leetcode/rust/src/435_non-overlapping_intervals.rs
pub fn erase_overlap_intervals(mut intervals: Vec<Vec<i32>>) -> i32 {
    intervals.sort();
    let mut res = 0;
    let mut prev = std::i32::MIN;
    for i in intervals {
        if i[0] >= prev {
            prev = i[1];
        } else {
            res += 1;
        }
    }
    res
}